thread-manager/src/tokio_runtime.rs

```diff
@@ -0,0 +1,178 @@
+    pub fn start_metrics_sampling(&self, period: Duration) {
+        let counters = self.counters.clone();
+        self.tokio.spawn(metrics_sampler(counters, period));
+    }
+
+    pub fn new(name: String, cfg: TokioConfig) -> anyhow::Result<Self> {
+        let num_workers = if cfg.worker_threads == 0 {
+            num_cpus::get()
+        } else {
+            cfg.worker_threads
+        };
+        let chosen_cores_mask = cfg.core_allocation.as_core_mask_vector();
+
+        let base_name = name.clone();
+        let mut builder = match num_workers {
+            1 => tokio::runtime::Builder::new_current_thread(),
+            _ => {
+                let mut builder = tokio::runtime::Builder::new_multi_thread();
+                builder.worker_threads(num_workers);
+                builder
+            }
+        };
+        let atomic_id: AtomicUsize = AtomicUsize::new(0);
+
+        let counters = Arc::new(ThreadCounters {
+            namespace: format!("thread-manager-tokio-{}", &base_name).leak(), // no workaround, metrics crate will only consume 'static str
+            parked_threads_cnt: AtomicI64::new(0),
+            active_threads_cnt: AtomicI64::new(
+                (num_workers.wrapping_add(cfg.max_blocking_threads)) as i64,
+            ),
+        });
+        let counters_clone1 = counters.clone();
+        let counters_clone2 = counters.clone();
+        builder
+            .event_interval(cfg.event_interval)
+            .thread_name_fn(move || {
+                let id = atomic_id.fetch_add(1, Ordering::Relaxed);
+                format!("{}-{}", base_name, id)
+            })
+            .on_thread_park(move || {
+                counters_clone1.on_park();
+            })
+            .on_thread_unpark(move || {
+                counters_clone2.on_unpark();
+            })
+            .thread_stack_size(cfg.stack_size_bytes)
+            .enable_all()
+            .max_blocking_threads(cfg.max_blocking_threads);
+
+        //keep borrow checker happy and move these things into the closure
+        let c = cfg.clone();
+        let chosen_cores_mask = Mutex::new(chosen_cores_mask);
+        builder.on_thread_start(move || {
+            let cur_thread = std::thread::current();
+            let _tid = cur_thread
+                .get_native_id()
+                .expect("Can not get thread id for newly created thread");
+            // todo - tracing
+            //let tname = cur_thread.name().unwrap();
+            //println!("thread {tname} id {tid} started");
+            apply_policy(
+                &c.core_allocation,
+                parse_policy(&c.policy),
+                c.priority,
+                &chosen_cores_mask,
+            );
+        });
+        Ok(TokioRuntime {
+            tokio: builder.build()?,
+            config: cfg.clone(),
+            counters,
+        })
+    }
+
+    /// Makes test runtime with 2 threads, only for unittests
+    pub fn new_for_tests() -> Self {
```

Comment:
`#[cfg(test)]`?

Reply:
No, can not be done as it is to be used from other crates.

Reply:
instead, you need to use
 
```rust
#[cfg(feature = "dev-context-only-utils")]
```

and to Cargo file
```rust
[features]
dev-context-only-utils = []
```

Reply:
Done!